primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1016 @@
+			vrf_inputs: impl IntoIterator<Item = VrfInput>,
+		) -> Self {
+			let vrf_inputs: Vec<VrfInput> = vrf_inputs.into_iter().collect();
+			let vrf_inputs = VrfIosVec::truncate_from(vrf_inputs);
+			let mut transcript = Transcript::new_labeled(label);
+			transcript_data
+				.into_iter()
+				.for_each(|data| transcript.append_slice(data.as_ref()));
+			VrfSignData { transcript, vrf_inputs }
+		}
+
+		/// Append a raw message to the transcript.
+		pub fn push_transcript_data(&mut self, data: &[u8]) {
+			self.transcript.append_slice(data);
+		}
+
+		/// Append a [`VrfInput`] to the vrf inputs to be signed.
+		///
+		/// On failure, gives back the [`VrfInput`] parameter.
+		pub fn push_vrf_input(&mut self, vrf_input: VrfInput) -> Result<(), VrfInput> {
+			self.vrf_inputs.try_push(vrf_input)
+		}
+
+		/// Create challenge from the transcript contained within the signing data.
+		pub fn challenge<const N: usize>(&self) -> [u8; N] {
+			let mut output = [0; N];
+			let mut transcript = self.transcript.clone();
+			let mut reader = transcript.challenge(b"Prehashed for bandersnatch");
+			reader.read_bytes(&mut output);
+			output
+		}
+	}
+
+	/// VRF signature.
+	#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct VrfSignature {
+		/// VRF (pre)outputs.
+		pub vrf_outputs: VrfIosVec<VrfOutput>,
+		/// VRF signature.
+		pub signature: Signature,
+	}
+
+	#[cfg(feature = "full_crypto")]
+	impl VrfCrypto for Pair {
+		type VrfInput = VrfInput;
+		type VrfOutput = VrfOutput;
+		type VrfSignData = VrfSignData;
+		type VrfSignature = VrfSignature;
+	}
+
+	#[cfg(feature = "full_crypto")]
+	impl VrfSecret for Pair {
+		fn vrf_sign(&self, data: &Self::VrfSignData) -> Self::VrfSignature {
+			// Workaround to overcome backend signature generic over the number of IOs.
+			match data.vrf_inputs.len() {
+				0 => self.vrf_sign_gen::<0>(data),
+				1 => self.vrf_sign_gen::<1>(data),
+				2 => self.vrf_sign_gen::<2>(data),
+				3 => self.vrf_sign_gen::<3>(data),
+				_ => unreachable!(),
+			}
+		}
+
+		fn vrf_output(&self, input: &Self::VrfInput) -> Self::VrfOutput {
+			let output = self.secret.0.vrf_preout(&input.0);
+			VrfOutput(output)
+		}
+	}
+
+	impl VrfCrypto for Public {
+		type VrfInput = VrfInput;
+		type VrfOutput = VrfOutput;
+		type VrfSignData = VrfSignData;
+		type VrfSignature = VrfSignature;
+	}
+
+	impl VrfPublic for Public {
+		fn vrf_verify(&self, data: &Self::VrfSignData, signature: &Self::VrfSignature) -> bool {
+			let preouts_len = signature.vrf_outputs.len();
+			if preouts_len != data.vrf_inputs.len() {
+				return false
+			}
+			// Workaround to overcome backend signature generic over the number of IOs.
+			match preouts_len {
+				0 => self.vrf_verify_gen::<0>(data, signature),
+				1 => self.vrf_verify_gen::<1>(data, signature),
+				2 => self.vrf_verify_gen::<2>(data, signature),
+				3 => self.vrf_verify_gen::<3>(data, signature),
+				_ => unreachable!(),
+			}
```

Comment:
So this is supposed to match `MAX_VRF_IOS`, right? In which case it'd be nice to have a static assert here that makes sure that `MAX_VRF_IOS == 3` (in case someone changes the constant in the future but forgets to update this function) Same for all other places where this is done.

Reply:
Good idea